src/tasks/task_list.hpp

```diff
@@ -85,8 +158,80 @@ class TaskList {
-                         [=]() mutable -> TaskStatus { return (obj->*func)(args...); });
+    return this->AddTask(dep, [=]() mutable -> TaskStatus {
+      return (obj->*func)(std::forward<Args>(args)...);
+    });
+  }
+
+  class IterativeTasks {
+   public:
+    IterativeTasks(TaskList *tl, int key) : tl_(tl), key_(key) {
+      max_iterations_ = std::numeric_limits<unsigned int>::max();
+    }
+
+    template <class T, class... Args>
+    TaskID AddTask(TaskID const &dep, T &&func, Args &&... args) {
+      return AddTask_(TaskType::iterative, dep, std::forward<T>(func),
+                      std::forward<Args>(args)...);
+    }
+    // overload to add member functions of class T to task list
+    // NOTE: we must capture the object pointer
+    template <class T, class... Args>
+    TaskID AddTask(TaskID const &dep, TaskStatus (T::*func)(Args...), T *obj,
+                   Args &&... args) {
+      return this->AddTask_(TaskType::iterative, dep, [=]() mutable -> TaskStatus {
+        return (obj->*func)(std::forward<Args>(args)...);
+      });
+    }
+
+    template <class T, class... Args>
+    TaskID AddCompletionTask(TaskID const &dep, T &&func, Args &&... args) {
+      return AddTask_(TaskType::completion_criteria, dep, std::forward<T>(func),
+                      std::forward<Args>(args)...);
+    }
+    template <class T, class... Args>
+    TaskID AddCompletionTask(TaskID const &dep, TaskStatus (T::*func)(Args...), T *obj,
+                             Args &&... args) {
+      return this->AddTask_(TaskType::completion_criteria, dep,
+                            [=]() mutable -> TaskStatus {
+                              return (obj->*func)(std::forward<Args>(args)...);
+                            });
+    }
+
+    void SetMaxIterations(const unsigned int max) { max_iterations_ = max; }
+    void SetFailWithMaxIterations(const bool flag) { throw_with_max_iters_ = flag; }
+    void SetWarnWithMaxIterations(const bool flag) { warn_with_max_iters_ = flag; }
+    bool ShouldThrowWithMax() const { return throw_with_max_iters_; }
+    bool ShouldWarnWithMax() const { return warn_with_max_iters_; }
+    unsigned int GetMaxIterations() const { return max_iterations_; }
+    unsigned int GetIterationCount() const { return count_; }
+    void IncrementCount() { count_++; }
+
+   private:
+    template <class F, class... Args>
+    TaskID AddTask_(const TaskType &type, TaskID const &dep, F &&func, Args &&... args) {
+      TaskID id(tl_->tasks_added_ + 1);
+      tl_->task_list_.push_back(Task(
+          id, dep,
+          [=, func = std::forward<F>(func)]() mutable -> TaskStatus {
+            return func(std::forward<Args>(args)...);
+          },
+          type, key_));
+      tl_->tasks_added_++;
+      return id;
+    }
+    TaskList *tl_;
+    int key_;
+    unsigned int max_iterations_;
+    unsigned int count_ = 0;
+    bool throw_with_max_iters_ = false;
+    bool warn_with_max_iters_ = true;
+  };
```

Comment:
I like this quite a bit... but I don't love that it's nested. Is there some way to not nest the classes?

Reply:
I don't know.  I did this because it allows these objects to directly access/modify private variables in the enclosing class (TaskLIst).  That was the most obvious thing to do, but there might be a better approach.

Reply:
Make the classes flat, but make them friends of each other. That should give private access.

Reply:
ok I separated the classes.  curious what you think

Reply:
I like it. I don't think it's too burdensome to return the task id in the way you do now. Seems entirely reasonable and clear how things are passed around.